### PR TITLE
feat(recommendations): change renderer implementation to VDOM

### PR DESCRIPTION
**Pull request by francoischalifour**
This is a rework of the rendering system of the Recommendations components to an agnostic virtual DOM implementation.

*Note that the Horizontal Slider UI component is not part of this PR, it'll be migrated in another stacked PR.*

## Problems

Here's a list of the problems we have with the current architecture:

- **Bundle size**. The JavaScript package embeds the complete `preact/compat` package which exports the whole React API. (We now only rely on `preact` and `preact/hooks`.)
- **Not future-proof**. The current React/Preact aliasing is not going to allow us to bring support for Vue, React Native, and all VDOM-based framework.
- **Bridge between JavaScript packages and UI components**. The bridge between UI components like Horizontal Slider and the Recommendations component is challenging because we're not in control of the `view` prop rendering framework.
- **Tooling**. The tooling is difficult to get working because we have to alias the dependencies in all our toolchain (Parcel, Jest, Rollup, Babel).

### Current

The previous packages main entry point was `react-recommendations`, where the logic and the rendering happened. The `js-recommendations` package relied on `react-recommendations` with an alias from `react` to `preact/compat`.

### New

We now have these packages:

```
packages
├── recommendations-core
├── recommendations-js
├── recommendations-react
└── recommendations-vdom
```

### `recommendations-core`

`recommendations-core` is meant to be self-deprecated once we support Recommend in the API client. It embeds the logic for fetching the recommendations, using the Algolia Search API.

**API**

- `getRecommendations`
- `getFrequentlyBoughtTogether`
- `getRelatedProducts`

### `recommendations-vdom`

`recommendations-vdom` is an agnostic package that allows to build components given a renderer. A renderer is composed of `{ createElement, Fragment }`. It's purely the UI part and creates uncontrolled components, lacking any state and lifecycle functionalities.

**API**

- `createFrequentlyBoughtTogetherComponent`
- `createRelatedProductsComponent`

**Usage**

```tsx
import { createElement, Fragment } from 'react';

const FrequentlyBoughtTogether = createFrequentlyBoughtTogetherComponent({
  createElement,
  Fragment,
});
```

### `recommendations-js`

`recommendations-js` is an implementation of `recommendations-vdom` with Preact (similar front-end framework that we use in Autocomplete and InstantSearch). It specifies `recommendations-vdom` with the Preact renderer.

**API**

- `relatedProducts`
- `frequentlyBoughtTogether`

### `recommendations-react`

`recommendations-react` is an implementation of `recommendations-vdom` with React.

**API**

- `RelatedProducts`
- `FrequentlyBoughtTogether`
- `useRecommendations`
- `useRelatedProducts`
- `useFrequentlyBoughtTogether`

### Other packages in the future

When the time comes, we'll be able to create `recommendations-vue` with the same strategy as `recommendations-js` and `recommendations-react`. We could also expose these components for React Native in the future.

## How to review

I'd recommend reading from package to package, and we can do a live code review if you'd like!

## Next steps

- Deprecate the previous packages
- If we decide to move forward with this new architecture, we'll send a stacked PR to also migrate the HorizontalSlider UI components.

### CODE REVIEW

**Comment by Haroenv** [comment]
Before reading the detailed code, I'm already letting know that I'm very excited about this layout of packages, it will solve a lot of problems we have now with mixing the preact/react in a single code base, making createElement & fragment opt-in is great (like in autocomplete) (in the mean time kinda thinking if it could make sense for InstantSearch, but I'm not really sure, since there the logic isn't just UI, but also clearly life cycle logic)

File: README.md
Comment (Haroenv) [inline README.md:14]: does -core and -vdom need to be added to this list?

File: bundlesize.config.json
Comment (Haroenv) [inline bundlesize.config.json:17]: why did the react package increase in size? Is that the overhead of the createXXX functions?

File: examples/js-demo/package.json
Comment (Haroenv) [inline examples/js-demo/package.json:12]: glad this isn't needed anymore

File: packages/recommendations-core/package.json
Comment (Haroenv) [inline packages/recommendations-core/package.json:2]: wow this diff is totally from the wrong file

File: packages/recommendations-core/src/__tests__/getRecommendations.test.ts
Comment (Haroenv) [inline packages/recommendations-core/src/__tests__/getRecommendations.test.ts:63]: why is this needed? maybe a comment?

File: packages/recommendations-core/src/getFrequentlyBoughtTogether.ts
Comment (Haroenv) [inline packages/recommendations-core/src/getFrequentlyBoughtTogether.ts:11]: this name feels a little bit like an unfinished sentence, since it's already implying e-commerce with the word "bought", what about "getFrequentlyBoughtTogetherProducts"?

File: packages/recommendations-core/src/getRecommendations.ts
Comment (Haroenv) [inline packages/recommendations-core/src/getRecommendations.ts:176]: should this not be an error instead of an empty state? It seems hard to debug

(sorry for commenting on unrelated code just because it renamed)

**Comment by francoischalifour** [comment]
@Haroenv This architecture has also sparkled some ideas for InstantSearch in my mind.

For instance, we could rely on a single `HierarchicalMenu` VDOM UI component that InstantSearch (either JS, React or Vue) can reuse and pass props to it, just like we're doing in this PR for `RelatedProducts`: https://github.com/algolia/ui-components/blob/a2ac8d5bbb68c044d931e29b53b1ca832d174d83/packages/recommendations-react/src/RelatedProducts.tsx

This means that we would be able to use `HierarchicalMenu` in the context of InstantSearch, but also elsewhere, as long as the props have the right interface. It'd bring really interesting use cases for Query Categorization and other new features, outside of strictly searching, but rather browsing/discovering.

The `useRelatedProducts` state logic would be our connector logic. Thinking about it this way, all the state and lifecycle logic would be part of InstantSearch, which would then populate the UI components. (It'll make sense once you'll see @shortcuts stacked PR for the HorizontalSlider component.)

InstantSearch would become just the search state manager with an API idiomatic to each framework, and it'd rely on agnostic UI components, shared among all frameworks. Plugging the life cycle logic isn't an unknown anymore as long as those components remain uncontrolled (i.e., they don't use any frameworks' specifics like `useState`).

That's just one application that I thought about when try these new patterns, but we can definitely discuss it further!

File: packages/recommendations-js/README.md
Comment (Haroenv) [inline packages/recommendations-js/README.md:10]: we'll need to make sure to `npm deprecate` the older package names once this is merged

File: packages/recommendations-js/src/frequentlyBoughtTogether.tsx
Comment (Haroenv) [inline packages/recommendations-js/src/frequentlyBoughtTogether.tsx]:
```suggestion
  const [result, setResult] = useState<GetRecommendationsResult<TObject>>({
```

I feel like this is named more naturally, since it's a network response/result, not really a return

File: packages/recommendations-react/src/useAlgoliaAgent.ts
Comment (Haroenv) [inline packages/recommendations-react/src/useAlgoliaAgent.ts:9]: in theory this code could also be shared between vdom implementations, but then you'll need to map to exactly useEffect in Vue, which might not be straightforward, so copied makes sense

File: packages/recommendations-vdom/README.md
Comment (Haroenv) [inline packages/recommendations-vdom/README.md:3]: I checked and react and preact both style it like this: VDOM, so it's good!

File: packages/recommendations-vdom/rollup.config.js
Comment (Haroenv) [inline packages/recommendations-vdom/rollup.config.js:9]: I wonder if the output could also be exported from the base rollup config (as a function that takes pkg), it's the same for all of them

File: packages/recommendations-vdom/src/DefaultFallback.tsx
Comment (Haroenv) [inline packages/recommendations-vdom/src/DefaultFallback.tsx:1]: it's a little confusing here because createElement isn't used as a prop. Should this maybe be inlined in where it's used, as it's not really a fallback, but nothing if it's not given

File: packages/recommendations-vdom/src/ListView.tsx
Comment (Haroenv) [inline packages/recommendations-vdom/src/ListView.tsx:33]: theres' another prop called props.Header (capitalised), maybe it's nicer if all components that are props are pascal case? IIRC that's a requirement if it's not read from an object directly, so being consistent can help

File: packages/recommendations-vdom/src/RelatedProducts.tsx
Comment (Haroenv) [inline packages/recommendations-vdom/src/RelatedProducts.tsx:51]: should this use the children-as-a-function pattern still? not sure if that translates to Vue efficiently

Comment (Haroenv) [reply packages/recommendations-vdom/src/RelatedProducts.tsx:51]: if children is a component, all those arguments could be props, not sure what the other options are

File: packages/recommendations-vdom/src/types/ProductRecord.ts
Comment (Haroenv) [inline packages/recommendations-vdom/src/types/ProductRecord.ts]: should this not be part of the -core package?

File: packages/recommendations-vdom/src/types/Renderer.ts
Comment (Haroenv) [inline packages/recommendations-vdom/src/types/Renderer.ts:5]: where does JSX get defined?

File: packages/recommendations-vdom/src/types/Renderer.ts
Comment (Haroenv) [inline packages/recommendations-vdom/src/types/Renderer.ts:9]: and this?

File: ship.config.js
Comment (Haroenv) [inline ship.config.js:10]: these packages get migrated in a different PR once this is approved?

**Review by Haroenv — COMMENTED** [review]
I love it!

File: ship.config.js
Comment (shortcuts) [reply ship.config.js:10]: The migration is already done on an other branch: https://github.com/algolia/ui-components/tree/feat/vdom-pt2 but we will open the second PR once we get some feedbacks on this implementation :D

File: README.md
Comment (shortcuts) [reply README.md:14]: I was also wondering, I've added them to the migration to VDOM part 2 but should we only state the "real" packages we provide and not the internal ones?

File: packages/recommendations-core/src/__tests__/getRecommendations.test.ts
Comment (shortcuts) [reply packages/recommendations-core/src/__tests__/getRecommendations.test.ts:63]: This might be because `createRecommendationsClient` creates a lite version of the `SearchClient` while `getRecommendations` doesn't?

File: packages/recommendations-vdom/rollup.config.js
Comment (shortcuts) [reply packages/recommendations-vdom/rollup.config.js:9]: Since we won't use rollup to build ESM I've inlined it in the part 2 of the migration, it's still different for the React packages since we use the `external` prop

File: packages/recommendations-vdom/src/types/Renderer.ts
Comment (shortcuts) [reply packages/recommendations-vdom/src/types/Renderer.ts:9]: https://github.com/algolia/ui-components/blob/a2ac8d5bbb68c044d931e29b53b1ca832d174d83/packages/recommendations-vdom/src/types/Renderer.ts#L18-L21

**Review by shortcuts — COMMENTED** [review]
So good!

File: README.md
Comment (francoischalifour) [reply README.md:14]: I was hesitant because those packages will probably remain internal, but that's fair. Added in 10272d48acf39bcd214e4cb7bd564a2e14cdb717.

File: bundlesize.config.json
Comment (francoischalifour) [reply bundlesize.config.json:17]: Good question! It went from 3.0KB to 3.1KB. Looking at the diff, it's likely because Babel transpile code multiple times (because the VDOM package is a new dependency) and those polyfills end up being duplicated in the bundle.

I'm able to counter-balance the cost of the abstraction and get it back down to 3.02KB by using the `@babel/plugin-transform-runtime` plugin in e0f94b0006000527f20a227c3877cd55734b5762. Lmk what you think?

File: packages/recommendations-core/src/__tests__/getRecommendations.test.ts
Comment (francoischalifour) [reply packages/recommendations-core/src/__tests__/getRecommendations.test.ts:63]: That's just temporary because the typings for the yet non-existant Recommendations API client don't exist. Since the expected search client and its typings are going to change very soon, I didn't bother hiding the typing issue.

File: packages/recommendations-core/src/getFrequentlyBoughtTogether.ts
Comment (francoischalifour) [reply packages/recommendations-core/src/getFrequentlyBoughtTogether.ts:11]: I agree—it would also be more consistent with "getRelated**Products**". The thing is it's been branded like that every where else on the AI team side. I'll catch up with them to see what they think about that.

File: packages/recommendations-core/src/getRecommendations.ts
Comment (francoischalifour) [reply packages/recommendations-core/src/getRecommendations.ts:176]: Since we mimic the final API with some workarounds with the Search endpoint, we didn't yet bother processing errors. I think we can rather tackle this more efficiently once we implement the new Recommendations API client spec, and have this discussion then. Wdyt?

File: packages/recommendations-js/src/frequentlyBoughtTogether.tsx
Comment (francoischalifour) [reply packages/recommendations-js/src/frequentlyBoughtTogether.tsx]: Addressed in 38f20bc3b64e1dd097b7148a284ba3a6b201c293.

File: packages/recommendations-react/src/useAlgoliaAgent.ts
Comment (francoischalifour) [reply packages/recommendations-react/src/useAlgoliaAgent.ts:9]: Yeah, here we want to avoid to use framework specifics and share them across packages, otherwise we go back to the original building/aliasing issues easily. This "code duplication" will be counter-balance with some testing.

File: packages/recommendations-vdom/rollup.config.js
Comment (francoischalifour) [reply packages/recommendations-vdom/rollup.config.js:9]: We plan to reiterate on the build scripts after @shortcuts stacked PR! (it'll include a fully-shared Rollup config, banners with license, etc.)

File: packages/recommendations-vdom/src/DefaultFallback.tsx
Comment (francoischalifour) [reply packages/recommendations-vdom/src/DefaultFallback.tsx:1]: I see it rather consistent with other default components. For now the default fallback is to not display anything but we're not sure if it'll stay like that.

I'd say let's keep it like that for now if you don't mind too much?

File: packages/recommendations-vdom/src/ListView.tsx
Comment (francoischalifour) [reply packages/recommendations-vdom/src/ListView.tsx:33]: I agree we need to find some consistency. We also have `fallbackComponent`. I think it's out of scope for this PR but let's definitely address this before public release.

File: packages/recommendations-vdom/src/RelatedProducts.tsx
Comment (francoischalifour) [reply packages/recommendations-vdom/src/RelatedProducts.tsx:51]: Having children as a function makes the API powerful for custom use cases, but the fact that it might not work with Vue is a good point. We'll need to test it before public release then. I guess converting that to a slot would work in the Vue API?

File: packages/recommendations-vdom/src/types/ProductRecord.ts
Comment (francoischalifour) [reply packages/recommendations-vdom/src/types/ProductRecord.ts]: Indeed—will move them.

File: packages/recommendations-vdom/src/types/Renderer.ts
Comment (francoischalifour) [reply packages/recommendations-vdom/src/types/Renderer.ts:5]: It's a TypeScript global, we also use it in Autocomplete.

File: packages/recommendations-vdom/src/types/Renderer.ts
Comment (francoischalifour) [reply packages/recommendations-vdom/src/types/Renderer.ts:9]: `VNode` is defined below.

File: packages/recommendations-js/README.md
Comment (francoischalifour) [reply packages/recommendations-js/README.md:10]: I added that point in the "Next steps" section—thanks!

File: packages/recommendations-vdom/src/types/ProductRecord.ts
Comment (francoischalifour) [reply packages/recommendations-vdom/src/types/ProductRecord.ts]: See 2fba5ac75b74a6d4ad6b591e153a5f79f203bfed.

File: README.md
Comment (Haroenv) [reply README.md:14]: could be in a separate header for internal packages?

File: packages/recommendations-vdom/src/ListView.tsx
Comment (Haroenv) [reply packages/recommendations-vdom/src/ListView.tsx:33]: I'll let you open an issue / note it down somewhere and this can be resolved

File: packages/recommendations-vdom/src/RelatedProducts.tsx
Comment (Haroenv) [reply packages/recommendations-vdom/src/RelatedProducts.tsx:51]: most likely, I didn't think of slots, so that should indeed solve the issue

**Review by Haroenv — APPROVED** [review]
I think for now this is good, although the remaining comments probably should move to issues eventually

File: packages/recommendations-vdom/src/ListView.tsx
Comment (francoischalifour) [reply packages/recommendations-vdom/src/ListView.tsx:33]: [Created a ticket on JIRA](https://algolia.atlassian.net/browse/DX-2016).

File: packages/recommendations-react/src/useAlgoliaAgent.ts
Comment (sarahdayan) [inline packages/recommendations-react/src/useAlgoliaAgent.ts]:
```suggestion
    props.searchClient.addAlgoliaAgent('recommendations-react', version);
```

**Review by sarahdayan — COMMENTED** [review]
Nice work, this layering feels a lot more future-proof. I'm also super excited to see how this can influence InstantSearch and maybe move the UI components outside, in this package 😍

@Haroenv and @shortcuts covered the PR and I agree with most comments, just added a minor one.

One question (unrelated to merging this PR)—with Autocomplete and now these UI components, it seems we're going with VDOM as the default for our vanilla JavaScript implementations. **Is it the direction we want to take? Are we also going to this route for InstantSearch.js?**

Personally I think it makes a lot of sense for performance and convenience concerns, and templating solutions don't come even close to what JSX expressions bring to the table. But this is also an opinionated choice which "force" users to use a VDOM implem and write JSX (or build VDOM trees by hand). I could see some users confused by this choice and I'm wondering if this is something we've vetted beforehand.

**Comment by Haroenv** [comment]
> Personally I think it makes a lot of sense for performance and convenience concerns, and templating solutions don't come even close to what JSX expressions bring to the table. But this is also an opinionated choice which "force" users to use a VDOM implem and write JSX (or build VDOM trees by hand). I could see some users confused by this choice and I'm wondering if this is something we've vetted beforehand.

I agree with your Sarah, For these smaller packages I think this makes sense, but I'm not fully convinced whether that's the right solutions for InstantSearch.js, templates (while having lots of problems) are a lot more approachable for a dev who's never used build systems or jsx

**Comment by sarahdayan** [comment]
@Haroenv Yeah, that's debatable though. When using a "small" package you might expect proportionate efforts, with less setup and a more plug-and-play approach.

There's also the concern of consistency, where developers using one of our UI solutions then move on to the next and there's a different approach to templating. Or even use them altogether and there are two ways of doing things.

**Comment by francoischalifour** [comment]
I've looked around quite a lot and for now I think that VDOM is the choice that **makes it easy to build interactive UIs** (which is the main point when dealing with search and discovery) both **for us as maintainers and for devs using our libs**.

(In InstantSearch, we rely on Preact internally, which uses VDOM, and the user-facing templating system is HTML/Hogan.js.)

I want to stress that what's interesting with this VDOM approach is that we can **offer users to easily bail out of JSX**, while still offering this VDOM layer. For example, we could detect that users are returning HTML strings and wrap that in a `createElement('div', { dangerouslySetInnerHTML: { __html: content }})`. Users could therefore **use Hogan.js if they need to avoid XSS**.

Example:

```tsx
relatedProducts({
  // ...
  itemComponent({ item }) {
    // We're able to detect HTML and forward it to `dangerouslySetInnerHTML`
    return `<h3>${item.name}</h3>`;
  },
});
```

What I'm trying to say is that **this abstraction is still permissive**, and it still allows users to use HTML if they want. If they express this need, we can create this layer to facilitate this alternative.

I also express some problems in solutions that don't provide diffing in the original Autocomplete VDOM PR: algolia/autocomplete#381. Accessibility is another concern that we already suffer from in InstantSearch.js.

As always, I'm open to any alternatives you've thought about.

File: packages/recommendations-react/src/useAlgoliaAgent.ts
Comment (francoischalifour) [reply packages/recommendations-react/src/useAlgoliaAgent.ts]: See 3c6e528ff1e233eb091b529c6386c15be789c0da.

**Review by shortcuts — APPROVED** [review]
🚀

**Comment by sarahdayan** [comment]
@francoischalifour Again, I'm fully sold on the benefits of using VDOM internally, and believe this is a sane pick for users as well. I think it's awesome if we can facilitate usage of template strings or a third-party template engine, and forward that ourselves to the `dangerouslySetInnerHTML` API because it caters to more users and avoid alienating those who have never seen JSX before.

Ultimately, users don't need to know we're using VDOM under the hood (just like with InstantSearch.js), it should be about what they want to use, and for this to feel natural.

**Comment by francoischalifour** [comment]
@sarahdayan I added "[Detect and forward HTML templates to `dangerouslySetInnerHTML`](https://algolia.atlassian.net/browse/DX-2085?atlOrigin=eyJpIjoiNTZhN2JlMjI4YTllNGYyMmI4ZTU3NTQ5MGJlYmRjNDciLCJwIjoiaiJ9)" to the backlog. It'll also be a nice addition to Autocomplete.

**Pull request merged**
